overlord/devicestate/devicestate.go

```diff
@@ -202,11 +203,63 @@ func checkGadgetOrKernel(st *state.State, snapInfo, curInfo *snap.Info, _ snap.C
 	return nil
 }
 
+var (
+	// TODO: replace with gadget.IsCompatible() once we are ready
+	gadgetIsCompatible = func(current, new *gadget.Info) error { return nil }
+)
+
+func checkGadgetRemodelCompatible(st *state.State, snapInfo, curInfo *snap.Info, snapf snap.Container, flags snapstate.Flags, deviceCtx snapstate.DeviceContext) error {
+	if release.OnClassic {
+		return nil
+	}
+	if snapInfo.GetType() != snap.TypeGadget {
+		// We are only interested in gadget snaps.
+		return nil
+	}
+	if deviceCtx == nil || !deviceCtx.ForRemodeling() {
+		// We are only interesting in a remodeling scenario.
+		return nil
+	}
+	if curInfo == nil {
+		// snap isn't installed yet, we are likely remodeling to a new
+		// gadget, identify the old gadget
+		model, err := findModel(st)
```

Comment:
no findModel  finds the ground/original model.  Anyway, because at some point model.Gadget() is going away (there will be only model.GadgetSnap()), it would be best to write this as:

```
		// snap isn't installed yet, we are likely remodeling to a new
		// gadget, identify the old gadget
		groundDeviceCtx, err := DeviceCtx(st, nil, nil)
		if err != nil || err == state.ErrNoState {
			return fmt.Errorf("cannot identify the current model")
		}
		curInfo, _ = snapstate.GadgetInfo(st, groundDeviceCtx)
```

Reply:
added